### PR TITLE
`duplicates()` now preserves groups and attributes for "regular" and "sf" data frames

### PR DESCRIPTION
**Pull request by dunkenwg**
There might be some problems with the way I coded this for sf data frames.
1. `duplicates()` now uses poisspatial which we'd have to include as a tidyplus package dependency.
2. When dealing with sf data frames, `duplicates()` first converts the `geometry` sfc column back to `X` and `Y` coordinate columns, then does `dplyr::inner_join()` (because `dplyr::inner_join()` doesn't work on sf data frames), then converts the coordinate columns back to the sfc column. Unfortunately, this all hinges on the sf data frame having an sfc column named `geometry` meaning the user can't change the name of the column before calling `duplicates()`. We could add another argument to `duplicates()` to resolve this issue but I'm worried this might clutter the function call.

Alternatively, we could use `sf::st_join()`. However, I don't think we can pass specific columns to `sf::st_join()` like we can with `dplyr::inner_join()`.

Let me know what you both think and I'll start writing tests for the new `duplicates()` functionality.

### CODE REVIEW

**Comment by joethorley** [comment]
sf data frames can have an active geometry column that doesn't have to have the same name as geometry. see 
`?sf::st_geometry`. There can also have more than one inactive geometry column.

The limitation of going to X and Y is that it only works for points. 

I think we either need to strip all geometry columns or find a more general solution. ie convert to text?

**Comment by dunkenwg** [comment]
`duplicates()` should now deal with any sf data frames. It deactivates sf data at the start with `tibble()` and reactivates it at the end with `ps_activate_sfc()`. Only downside is we'd still need poisspatial as a dependency. Let me know if this is what you were looking for.

File: R/duplicates.R [hunk 1]
```diff
@@ -21,26 +21,30 @@ duplicates <- function(.data, ..., .keep_all = TRUE) {
   chk_flag(.keep_all)
 
   col <- rlang::ensyms(...)
-  if (length(col) == 0) {
-    col_names <- colnames(.data)
-  } else {
-    col_names <- vapply(col, rlang::as_string, character(1))
-  }
+  if (length(col) == 0) { col_names <- colnames(.data) }
```
Comment (aylapear) [inline R/duplicates.R]: I'd format this how you had it before.

File: R/duplicates.R [hunk 2]
```diff
@@ -21,26 +21,30 @@ duplicates <- function(.data, ..., .keep_all = TRUE) {
...
+  else { col_names <- vapply(col, rlang::as_string, character(1)) }
   col_names <- unique(col_names)
   chk_vector(col_names)
   check_values(col_names, "")
   check_names(.data, col_names)
 
-  if (!length(col_names)) {
-    return(.data)
-  }
+  if (!length(col_names)) { return(.data) }
```
Comment (aylapear) [inline R/duplicates.R]: I'd have this formatted how you had it before.

File: R/duplicates.R [hunk 3]
```diff
@@ -21,26 +21,30 @@ duplicates <- function(.data, ..., .keep_all = TRUE) {
...
+  
+  grouped <- dplyr::is_grouped_df(.data)
+  groups <- dplyr::group_vars(.data)
+  groups_sym <- rlang::syms(groups)
+  
+  is_sf = any(class(.data) == "sf")
```
Comment (aylapear) [inline R/duplicates.R]: Use the assignment arrow here instead of the equals symbol

File: R/duplicates.R [hunk 4]
```diff
@@ -21,26 +21,30 @@ duplicates <- function(.data, ..., .keep_all = TRUE) {
...
-  .data <- dplyr::inner_join(.data, .data_dup, by = col_names)
-  if (!(.keep_all)) {
-    .data <- dplyr::select(.data, dplyr::all_of(col_names))
-  }
-  .data <- dplyr::as_tibble(.data)
+  
+  .data <- dplyr::inner_join(.data, .data_dup, by = col_names) 
+  if (!(.keep_all)) { .data <- dplyr::select(.data, dplyr::all_of(col_names)) }
+  if (grouped) { .data <- dplyr::group_by(.data, !!!groups_sym) }
+  if (is_sf) { .data <- poisspatial::ps_activate_sfc(.data) }
```
Comment (aylapear) [inline R/duplicates.R]: I am not sure if we should set the active geometry here as it depends on the column name being "geometry".

Comment (joethorley) [reply R/duplicates.R:48]: Yes I think we should the following functions to get, save and then set the active geometry name

https://r-spatial.github.io/sf/reference/st_geometry.html

**Comment by aylapear** [comment]
I am leaning towards that we just set the table to be a tibble and not re-active the spatial table afterwards. If we go this direction we need to state it in the function details. 

If the geometry is active when it goes into the function does it affect the outcome?

**Comment by joethorley** [comment]
Yes if a geometry is active the object behaves very different (basically a sf as opposed to just a tibble) but it is possible to (re)active the active geometry column by name  and to get the name of the active geometry column at the start.

https://r-spatial.github.io/sf/reference/st_geometry.html

**Comment by joethorley** [comment]
I just did a push to the branch of a test that preserves single active geometry column called geometry (its currently failing)

**Comment by dunkenwg** [comment]
I made some minor edits to your test and everything looks like it's working properly. I also added another test to make sure that `duplicates()` works on grouped data frames and sf columns with unconventional names. If these tests look good to you, I'll write a couple more to really make sure that `duplicates()` works on groups and sf columns.

**Comment by joethorley** [comment]
We can't use poisspatial as a dependency in tidyplus as poisspatial is not a CRAN package. Ultimately poisspatial is just a wrapper on sf functions so you can activate using functions in

[r-spatial.github.io/sf/reference/st_geometry.html](https://r-spatial.github.io/sf/reference/st_geometry.html)

**Comment by dunkenwg** [comment]
Hey @joethorley, I've removed poisspatial and split the tests up for groups and sf data frames. Let me know what you think and if you want me to add any more tests or start looking through the other functions in tidyplus.

File: tests/testthat/test-duplicates.R [hunk 5]
```diff
@@ -106,3 +96,67 @@ test_that("errors when input argument is not a data.frame", {
...
+  
+  data_dup <- tibble::tibble(
+    X = c(3, 3, 4, 4),
+    Y = c(14, 14, 15, 15),
+    a = c("green", "green", "blue", "blue"),
+    b = c("white", "white", "white", "white")
+  )
+  data_dup <- dplyr::group_by(data_dup, a, b)
+  
+  expect_identical(data_dup, duplicates(data))
```
Comment (aylapear) [inline tests/testthat/test-duplicates.R]: I would apply the `duplicates()` function outside of the expect call if possible as it makes the code more readable.

File: tests/testthat/test-duplicates.R [hunk 6]
```diff
@@ -106,3 +96,67 @@ test_that("errors when input argument is not a data.frame", {
...
+  
+  data_dup <- tibble::tibble(
+    X = c(3, 3, 4, 4),
+    Y = c(14, 14, 15, 15),
+    a = c("green", "green", "blue", "blue"),
+    b = c("white", "white", "white", "white")
+  )
+  data_dup <- sf::st_as_sf(data_dup, coords = c("X", "Y"))
+
+  expect_identical(data_dup, duplicates(data))
```
Comment (aylapear) [inline tests/testthat/test-duplicates.R]: Same as comment above.

File: tests/testthat/test-duplicates.R [hunk 7]
```diff
@@ -106,3 +96,67 @@ test_that("errors when input argument is not a data.frame", {
...
+  
+  data_dup <- tibble::tibble(
+    X = c(3, 3, 4, 4),
+    Y = c(14, 14, 15, 15),
+    a = c("green", "green", "blue", "blue"),
+    b = c("white", "white", "white", "white")
+  )
+  data_dup <- sf::st_as_sf(data_dup, coords = c("X", "Y"), sf_column_name = "map")
+  
+  expect_identical(data_dup, duplicates(data))
```
Comment (aylapear) [inline tests/testthat/test-duplicates.R]: Same as above

**Review by aylapear — CHANGES_REQUESTED** [review]
I think it would be worth adding a test for if you have more then 1 sf column in the same data frame, so you would have 1 active and 1 non-active spatial column. This would also cover off on the case for how the function handles non-active spatial columns.

**Comment by aylapear** [comment]
You should read this section about packages that go in Suggests for CRAN packages. https://poisson-software-dev.netlify.app/cran-release#uncommon-issues as `sf` falls under this and we just learned about this from another packages that was recently submitted to CRAN.

**Comment by aylapear** [comment]
Associated with this issue #20

File: tests/testthat/test-duplicates.R (hunk 5 above)
Comment (dunkenwg) [reply tests/testthat/test-duplicates.R]: @aylapear, I did it this way to be consistent with the previously written tests. Would you like me to change all of the tests, just the tests that I wrote, or keep it the way it is now?

Comment (aylapear) [reply tests/testthat/test-duplicates.R]: Just the tests you wrote for now.

**Review by aylapear — CHANGES_REQUESTED** [review]

**Review by joethorley — APPROVED** [review]

**Review by aylapear — APPROVED** [review]

**Comment by aylapear** [comment]
@dunkenwg Changes all look good, can you use the new fledge trick we learned to update the version a dev bump and the news file. I will merge it in since it is a CRAN package.

**Comment by dunkenwg** [comment]
Resolves #20

File: DESCRIPTION [hunk 8]
```diff
@@ -1,6 +1,6 @@
 Package: tidyplus
 Title: Additional 'tidyverse' Functions
-Version: 0.1.0
+Version: 0.1.0.9000
```
Comment (aylapear) [inline DESCRIPTION:3]:
```suggestion
Version: 0.1.0.9001
```

Comment (aylapear) [reply DESCRIPTION:3]: This was a test to practice the commit suggestion thing and thus can be rejected.

**Review by aylapear — APPROVED** [review]
